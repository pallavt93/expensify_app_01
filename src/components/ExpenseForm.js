import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

console.log(now.format());

export default class ExpenseForm extends React.Component{

    constructor(props){
        super(props);

        this.state = {
            description: props.expense ? props.expense.description: "",
            note: props.expense ? props.expense.note: "",
            amount: props.expense ? props.expense.amount.toString(): "",
            createdAt: props.expense ? moment(props.expense.createdAt): moment(),
            calenderFocused: false,
            error: ""
        };
    }

    // state = {
    //     description: "",
    //     note: "",
    //     amount:"",
    //     createdAt: moment(),
    //     calenderFocused: false,
    //     error: ""
    // };

    onChangeDescription = (e)=>{
        const description = e.target.value;
        this.setState(()=>({ description }));
    };

    onChangeNote = (e)=>{
        const note = e.target.value;
        this.setState(()=>({ note }));
    };

    onChangeAmount = (e)=>{
        const amount = e.target.value;
        if (!amount || amount.match( /^\d{1,}(\.\d{0,2})?$/) ){
            this.setState(()=>({ amount }));
        }
        
    };

    onDateChange = (createdAt)=>{
        if(createdAt)
        {
            this.setState(()=>({ createdAt }));
        }
        
    };

    onFocusChange = ({ focused })=>{
        this.setState(()=>({ calenderFocused: focused }))
    };

    onSubmit = (e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error: 'Please add valid description and amount'}));
        }
        else{
            this.setState(()=>({error: ''}));
            console.log("this is working...");
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="enter description" autoFocus
                        value={this.state.description} onChange={this.onChangeDescription}
                    />

                    <input type="text" placeholder="enter amount"
                        value={this.state.amount} onChange={this.onChangeAmount} 
                    />

                    <SingleDatePicker 
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calenderFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>{false}}
                    />

                    <textarea placeholder="add a note for your expense"
                        value={this.state.note} onChange={this.onChangeNote}
                    >

                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}