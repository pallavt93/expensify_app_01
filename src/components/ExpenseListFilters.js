import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate,setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component{
    state = {
        calenderFocused:null
    };

    onDatesChange = ({startDate,endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calenderFocused)=>{
        this.setState(()=>({ calenderFocused }));
    };

    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e)=>{
        const sortBy = e.target.value;
        if(sortBy === 'date') this.props.sortByDate();
        else if(sortBy === 'amount') this.props.sortByAmount()
    };

    render(){
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate= {this.props.filters.endDate}
                    onDatesChange = {this.onDatesChange}
                    focusedInput = {this.state.calenderFocused}
                    onFocusChange = {this.onFocusChange}
                    showClearDates = {true}
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                />
            </div>
        ); 
    }
}

// const ExpenseListFilters = (props)=>(
//     <div>
//         <input type="text" value={props.filters.text} onChange={(e)=>{
//             props.dispatch(setTextFilter(e.target.value));
//         }}/>
//         <select value={props.filters.sortBy} onChange={(e)=>{
//             const sortBy = e.target.value;
//             if(sortBy === 'date') props.dispatch(sortByDate());
//             else if(sortBy === 'amount') props.dispatch(sortByAmount())
//         }}>
//             <option value='date'>Date</option>
//             <option value='amount'>Amount</option>
//         </select>
//     </div>
// ); 

const mapStatetoProps = (state)=>{
    return {
        filters: state.filters
    };
};

const mapDisPatchToProps = (dispatch)=>({
    setTextFilter: (text)=>dispatch(setTextFilter(text)),
    sortByDate: ()=>dispatch(sortByDate()),
    sortByAmount: ()=>dispatch(sortByAmount()),
    setStartDate: (startDate)=>dispatch(setStartDate(startDate)),
    setEndDate: (endDate)=>dispatch(setEndDate(endDate))
});

export default connect(mapStatetoProps, mapDisPatchToProps)(ExpenseListFilters);