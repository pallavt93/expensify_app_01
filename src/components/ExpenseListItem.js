import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt })=>(
        <Link className="list-item" to={`/edit/${id}`}>
            <div>
                <h3 className="list-item__title">{description}</h3>
                <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
            </div>
            <h3 className="list-item__data">{numeral(amount).format('$0,0.00')}</h3> 
        </Link>
);

export default ExpenseListItem;

// import React from 'react';
// import { removeExpense } from '../actions/expenses';
// import { connect } from 'react-redux';
// import { Link } from 'react-router-dom';

// const ExpenseListItem = ({ dispatch, id, description, amount, createdAt })=>(
//     <div>
//         <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
//         <p>{amount} {createdAt}</p>
//         <button onClick = {()=>{ 
//             dispatch(removeExpense({ id }));
//         }}>remove</button>
//     </div>
// );

// export default connect()(ExpenseListItem);;