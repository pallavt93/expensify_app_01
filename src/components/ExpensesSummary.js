import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses_total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';


export const ExpensesSummary = ({expensesCount, expensesTotal})=>{
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const formattedAmount = numeral(expensesTotal).format('$0,0.00')
    return (
        <div>
            <h3>Viewing {expensesCount} {expenseWord} totaling {formattedAmount}.</h3>
        </div>
    );
};


const mapStateToProps = (state)=>{
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);