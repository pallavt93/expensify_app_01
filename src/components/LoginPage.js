import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage =  ( { startLogin }) => (
    <div>
        LoginPage
        <div>
            <button onClick={startLogin}>log in</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch)=>({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);