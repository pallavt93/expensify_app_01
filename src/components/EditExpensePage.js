import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense  } from '../actions/expenses';


export class EditExpensePage extends React.Component{
    onSubmit = (expense)=>{
        this.props.editExpense(expense);
        this.props.history.push('/');
    };

    onClick = ()=>{
        this.props.removeExpense(); 
        console.log(this.props.expense);
        this.props.history.push('/');
    };


    render(){
        return (
            <div>
                <ExpenseForm
                    expense = {this.props.expense} 
                    onSubmit = {this.onSubmit}
                />
                <div>
                <button 
                    onClick = {this.onClick}
                >remove
                </button>
                </div>
            </div>
        );
    }
};

// const EditExpensePage = (props) => {
//     return (
//         <div>
//             <ExpenseForm
//                 expense = {props.expense} 
//                 onSubmit = {(expense)=>{
//                     console.log(expense);
//                     props.dispatch(editExpense(props.match.params.id,expense));
//                     props.history.push('/');
//                 }}
//             />
//             <div>
//             <button 
//                 onClick = {
//                             ()=>{ props.dispatch(removeExpense({ id:props.expense.id })); 
//                                     props.history.push('/');
//                                  }
//                             }
//             >remove
//             </button>
//             </div>
//         </div>
//     );
// }

const mapDispatchToProps = (dispatch,props)=>({
    editExpense: (expense)=>dispatch(editExpense(props.match.params.id,expense)),
    removeExpense:()=>dispatch(removeExpense({ id:props.match.params.id}))
});

const mapStateToProps = (state, props)=>{
    return {
        expense: state.expenses.find((expense)=> expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
//export default connect(mapStateToProps)(EditExpensePage);