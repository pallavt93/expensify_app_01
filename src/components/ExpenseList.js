import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ?
            (
                <p>No Expenses</p>
            )
            : 
            (
                props.expenses.map((expense)=>( <ExpenseListItem { ...expense }  key={expense.id} />) )
            )
        }
    </div>
);

const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);




/*
const ConnectedExpenseList = connect((state)=>{
    return {
        expenses: state.expenses
    };
})
(ExpenseList);

export default ConnectedExpenseList;
*/