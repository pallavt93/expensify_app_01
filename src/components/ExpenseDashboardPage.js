import React from 'react';
import { connect } from 'react-redux';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;