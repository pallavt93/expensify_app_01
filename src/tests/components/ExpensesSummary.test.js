import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render Expenses Summary correctly with one expense',()=>{
    const wrapper  = shallow(<ExpensesSummary expensesCount={1} expensesTotal={100}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render Expenses Summary correctly with multiple expenses',()=>{
    const wrapper  = shallow(<ExpensesSummary expensesCount={4} expensesTotal={500} />);
    expect(wrapper).toMatchSnapshot();
});