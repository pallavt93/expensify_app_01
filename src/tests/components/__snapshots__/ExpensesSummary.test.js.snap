// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expenses Summary correctly with multiple expenses 1`] = `
<div>
  <h3>
    Viewing 
    4
     
    expenses
     totaling 
    $500.00
    .
  </h3>
</div>
`;

exports[`should render Expenses Summary correctly with one expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totaling 
    $100.00
    .
  </h3>
</div>
`;
